# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** [RaShaun Jackson]

**Student ID:** [Enter Your Student ID]

**Date Submitted:** [10/11/2025]

**Due Date:** [10/12/2025

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [4]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)

# TODO: Load lubridate
library(lubridate)

cat("✅ Packages loaded successfully!\n")


✅ Packages loaded successfully!


In [9]:
# ===================================================
# Load Required Packages
# ===================================================
library(tidyverse)
library(lubridate)

# ===================================================
# Set Working Directory
# ===================================================
setwd("/workspaces/Fall2025-MS3083-Base_Template/data")

# ===================================================
# Import Datasets
# ===================================================
feedback <- read_csv("customer_feedback.csv")
transactions <- read_csv("transaction_log.csv")
products <- read_csv("product_catalog.csv")

# ===================================================
# Display Data Structure and Samples
# ===================================================
cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n\n")

# Show structure of each dataset
cat("=== STRUCTURE: CUSTOMER FEEDBACK ===\n")
print(glimpse(feedback))
cat("\n=== STRUCTURE: TRANSACTIONS ===\n")
print(glimpse(transactions))
cat("\n=== STRUCTURE: PRODUCTS ===\n")
print(glimpse(products))

# Show first few rows for a quick view
cat("\n=== SAMPLE ROWS: CUSTOMER FEEDBACK ===\n")
print(head(feedback, 5))
cat("\n=== SAMPLE ROWS: TRANSACTIONS ===\n")
print(head(transactions, 5))
cat("\n=== SAMPLE ROWS: PRODUCTS ===\n")
print(head(products, 5))



Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (2): Feedback_Text, Contact_Info
dbl  (2): FeedbackID, CustomerID
date (1): Feedback_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 150 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): Transaction_DateTime, Status
dbl (3): LogID, CustomerID, Amount

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 75 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): Product_Description, Category, In_Stock
dbl (2): ProductID, Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column typ

✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 150 
Product rows: 75 

=== STRUCTURE: CUSTOMER FEEDBACK ===
Rows: 100
Columns: 5
$ FeedbackID    <dbl> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
$ CustomerID    <dbl> 12, 40, 34, 1, 47, 13, 13, 37, 49, 23, 24, 35, 3, 20, 38…
$ Feedback_Text <chr> "Highly recommend this item", "Excellent service", "Poor…
$ Contact_Info  <chr> "bob.wilson@test.org", "555-123-4567", "jane_smith@compa…
$ Feedback_Date <date> 2024-02-23, 2024-01-21, 2023-09-02, 2023-08-21, 2023-04…
# A tibble: 100 × 5
   FeedbackID CustomerID Feedback_Text                Contact_Info Feedback_Date
        <dbl>      <dbl> <chr>                        <chr>        <date>       
 1          1         12 Highly recommend this item   bob.wilson@… 2024-02-23   
 2          2         40 Excellent service            555-123-4567 2024-01-21   
 3          3         34 Poor quality control         jane_smith@… 2023-09-02   
 4          4          1 

In [10]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# Display structure of feedback using str()
str(feedback)

# Display first 5 rows of feedback
head(feedback, 5)

cat("\n=== TRANSACTION DATA ===\n")
# Display structure of transactions
str(transactions)

# Display first 5 rows of transactions
head(transactions, 5)

cat("\n=== PRODUCT CATALOG DATA ===\n")
# Display structure of products
str(products)

# Display first 5 rows of products
head(products, 5)


=== CUSTOMER FEEDBACK DATA ===
spc_tbl_ [100 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ FeedbackID   : num [1:100] 1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : num [1:100] 12 40 34 1 47 13 13 37 49 23 ...
 $ Feedback_Text: chr [1:100] "Highly recommend this item" "Excellent service" "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr [1:100] "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: Date[1:100], format: "2024-02-23" "2024-01-21" ...
 - attr(*, "spec")=
  .. cols(
  ..   FeedbackID = col_double(),
  ..   CustomerID = col_double(),
  ..   Feedback_Text = col_character(),
  ..   Contact_Info = col_character(),
  ..   Feedback_Date = col_date(format = "")
  .. )
 - attr(*, "problems")=<externalptr> 


FeedbackID CustomerID Feedback_Text                    Contact_Info          
1 1          12         Highly recommend this item       bob.wilson@test.org   
2 2          40         Excellent service                555-123-4567          
3 3          34         Poor quality control             jane_smith@company.com
4 4           1         average product, nothing special jane_smith@company.com
5 5          47         AMAZING customer support!!!      555-123-4567          
  Feedback_Date
1 2024-02-23   
2 2024-01-21   
3 2023-09-02   
4 2023-08-21   
5 2023-04-24


=== TRANSACTION DATA ===
spc_tbl_ [150 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ LogID               : num [1:150] 1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID          : num [1:150] 26 21 12 6 32 27 31 30 31 13 ...
 $ Transaction_DateTime: chr [1:150] "4/5/24 14:30" "3/15/24 14:30" "3/15/24 14:30" "3/20/24 9:15" ...
 $ Amount              : num [1:150] 277 175 252 215 269 ...
 $ Status              : chr [1:150] "Pending" "Pending" "Pending" "Pending" ...
 - attr(*, "spec")=
  .. cols(
  ..   LogID = col_double(),
  ..   CustomerID = col_double(),
  ..   Transaction_DateTime = col_character(),
  ..   Amount = col_double(),
  ..   Status = col_character()
  .. )
 - attr(*, "problems")=<externalptr> 


LogID CustomerID Transaction_DateTime Amount Status   
1 1     26         4/5/24 14:30         277.22 Pending  
2 2     21         3/15/24 14:30        175.16 Pending  
3 3     12         3/15/24 14:30        251.71 Pending  
4 4      6         3/20/24 9:15         214.98 Pending  
5 5     32         3/20/24 9:15         268.91 Completed


=== PRODUCT CATALOG DATA ===
spc_tbl_ [75 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ ProductID          : num [1:75] 1 2 3 4 5 6 7 8 9 10 ...
 $ Product_Description: chr [1:75] "Apple iPhone 14 Pro - 128GB - Space Black" "samsung galaxy s23 ultra 256gb" "Apple iPhone 14 Pro - 128GB - Space Black" "Apple iPhone 14 Pro - 128GB - Space Black" ...
 $ Category           : chr [1:75] "TV" "TV" "Audio" "Shoes" ...
 $ Price              : num [1:75] 964 1817 853 649 586 ...
 $ In_Stock           : chr [1:75] "Limited" "Yes" "Yes" "Yes" ...
 - attr(*, "spec")=
  .. cols(
  ..   ProductID = col_double(),
  ..   Product_Description = col_character(),
  ..   Category = col_character(),
  ..   Price = col_double(),
  ..   In_Stock = col_character()
  .. )
 - attr(*, "problems")=<externalptr> 


ProductID Product_Description                       Category    Price  
1 1         Apple iPhone 14 Pro - 128GB - Space Black TV           963.53
2 2         samsung galaxy s23 ultra 256gb            TV          1817.44
3 3         Apple iPhone 14 Pro - 128GB - Space Black Audio        852.79
4 4         Apple iPhone 14 Pro - 128GB - Space Black Shoes        648.58
5 5         samsung galaxy s23 ultra 256gb            Electronics  586.35
  In_Stock
1 Limited 
2 Yes     
3 Yes     
4 Yes     
5 Limited

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [14]:
# ===================================================
# Task 2.1: Clean Product Names
# ===================================================
library(tidyverse)

# --- Step 1: Detect the most likely product name column automatically
name_candidates <- c(
  "product_name", "Product_Name", "productname", "ProductName",
  "name", "Name", "product", "Product", "product_title", "Product_Title",
  "Description", "Product_Description"
)

found_name_col <- intersect(name_candidates, names(products))[1]

if (is.na(found_name_col)) {
  cat("⚠️ Could not automatically find a product name column. Please run colnames(products) to verify.\n")
} else {
  cat("✅ Found product name column:", found_name_col, "\n")

  # --- Step 2: Create a clean version
  products_clean <- products %>%
    mutate(
      product_name_clean = !!sym(found_name_col) %>%
        stringr::str_trim() %>%      # remove leading/trailing spaces
        stringr::str_squish() %>%    # remove extra internal spaces
        stringr::str_to_title()      # convert to title case
    )

  # --- Step 3: Display before/after for inspection
  cat("\nProduct Name Cleaning Results:\n")
  products_clean %>%
    select(all_of(found_name_col), product_name_clean) %>%
    head(10) %>%
    print()
}



✅ Found product name column: Product_Description 

Product Name Cleaning Results:
# A tibble: 10 × 2
   Product_Description                         product_name_clean               
   <chr>                                       <chr>                            
 1 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 2 "samsung galaxy s23 ultra 256gb"            "Samsung Galaxy S23 Ultra 256gb" 
 3 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 4 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 5 "samsung galaxy s23 ultra 256gb"            "Samsung Galaxy S23 Ultra 256gb" 
 6 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 7 "DELL XPS 13 Laptop - Intel i7 - 16GB RAM"  "Dell Xps 13 Laptop - Intel I7 -…
 8 "hp envy printer - wireless - color"        "Hp Envy Printer - Wireless - Co…
 9 "Nike Air Max 270 - Size 10 - Black/White"  "Nike Air Max 270 - Size 10 - Bl…
10 "LG 5

In [16]:
# --- Quick validation & diagnostics ---

# 1) Confirm required columns exist
stopifnot(all(c("OrderID","CustomerID") %in% names(orders_items)))
stopifnot(all(c("CustomerID") %in% names(orders_customers_items)))
stopifnot(all(c("ProductID","Supplier_ID") %in% names(complete_order_data)))
stopifnot(all(c("Supplier_ID") %in% names(suppliers)))

# 2) Show shapes
cat("customers:", nrow(customers), "rows | orders:", nrow(orders), 
    "| order_items:", nrow(order_items), "| products:", nrow(products),
    "| suppliers:", nrow(suppliers), "\n")
cat("orders_items:", nrow(orders_items), 
    "| orders_customers_items:", nrow(orders_customers_items),
    "| complete_order_data:", nrow(complete_order_data),
    "| complete_data:", nrow(complete_data), "\n")

# 3) Check for unmatched suppliers (orders referencing suppliers not in suppliers table)
orders_missing_suppliers <- anti_join(complete_order_data, suppliers, by = "Supplier_ID")
cat("Orders with missing suppliers:", nrow(orders_missing_suppliers), "\n")

# 4) Check for suppliers never used
unused_suppliers <- anti_join(suppliers, complete_order_data, by = "Supplier_ID")
cat("Suppliers with no orders/products:", nrow(unused_suppliers), "\n")

# 5) Duplicate key checks (common cause of row explosions)
dup_products  <- sum(duplicated(products$ProductID))
dup_suppliers <- sum(duplicated(suppliers$Supplier_ID))
cat("Duplicate ProductID:", dup_products, "| Duplicate Supplier_ID:", dup_suppliers, "\n")

# 6) Peek at final dataset
cat("\n=== complete_data preview ===\n")
print(head(complete_data, 5))



customers: 100 rows | orders: 250 | order_items: 400 | products: 50 | suppliers: 10 
orders_items: 400 | orders_customers_items: 310 | complete_order_data: 310 | complete_data: 310 


Orders with missing suppliers: 0 
Suppliers with no orders/products: 0 
Duplicate ProductID: 0 | Duplicate Supplier_ID: 0 

=== complete_data preview ===
# A tibble: 5 × 16
  CustomerID Name  Email City  Registration_Date OrderID Order_Date Total_Amount
       <dbl> <chr> <chr> <chr> <date>              <dbl> <date>            <dbl>
1          1 Cust… cust… Phoe… 2020-10-03             87 2023-03-28         716.
2          1 Cust… cust… Phoe… 2020-10-03            214 2023-09-12        1344.
3          1 Cust… cust… Phoe… 2020-10-03            214 2023-09-12        1344.
4          1 Cust… cust… Phoe… 2020-10-03            214 2023-09-12        1344.
5          3 Cust… cust… Chic… 2021-04-20             29 2023-03-07         441.
# ℹ 8 more variables: ProductID <dbl>, Quantity <dbl>, Unit_Price <dbl>,
#   Product_Name <chr>, Category <chr>, Supplier_ID <dbl>, Supplier_Name <chr>,
#   Country <chr>


In [21]:
# Task 2.2: Standardize Product Categories (robust to different column names)
library(tidyverse)

# If Task 2.1 hasn't run, fall back to raw products
if (!exists("products_clean")) products_clean <- products

# 1) Detect the category column automatically
category_candidates <- c(
  "category","Category","product_category","Product_Category",
  "category_name","CategoryName","department","Department",
  "segment","Segment","class","Class"
)
found_category_col <- intersect(category_candidates, names(products_clean))[1]

if (is.na(found_category_col)) {
  stop("No category-like column found in products/products_clean. ",
       "Available columns: ", paste(names(products_clean), collapse = ", "))
}

# 2) Create standardized category column
products_clean <- products_clean %>%
  mutate(
    category_clean = .data[[found_category_col]] %>%
      stringr::str_squish() %>%     # remove extra internal + leading/trailing spaces
      stringr::str_to_title()       # Title Case
  )

# 3) Show unique categories before and after
cat("Original categories (from column: ", found_category_col, "):\n", sep = "")
print(unique(products_clean[[found_category_col]]))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))


Original categories (from column: Category):


[1] "TV"          "Audio"       "Shoes"       "Electronics" "Computers"  

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [23]:
# Task 2.3: Clean Customer Feedback Text (robust to different column names)
library(tidyverse)

# Detect the feedback text column automatically
feedback_text_candidates <- c(
  "feedback_text","Feedback_Text","feedback","Feedback",
  "comments","Comments","comment","Comment",
  "review","Review","message","Message","text","Text"
)
fb_col <- intersect(feedback_text_candidates, names(feedback))[1]

if (is.na(fb_col)) {
  stop("No feedback text column found. Available columns: ",
       paste(names(feedback), collapse = ", "))
}

# Create 'feedback_clean': lowercase + squish whitespace
feedback_clean <- feedback %>%
  mutate(
    feedback_clean = .data[[fb_col]] %>%
      stringr::str_to_lower() %>%
      stringr::str_squish()
  )

# Display sample (show original detected column alongside the cleaned text)
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
  transmute(!!fb_col := .data[[fb_col]], feedback_clean) %>%
  head(5) %>%
  print()



Feedback Cleaning Sample:
# A tibble: 5 × 2
  Feedback_Text                    feedback_clean                  
  <chr>                            <chr>                           
1 Highly recommend this item       highly recommend this item      
2 Excellent service                excellent service               
3 Poor quality control             poor quality control            
4 average product, nothing special average product, nothing special
5 AMAZING customer support!!!      amazing customer support!!!     


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [26]:
# ===================================================
# Task 3.1: Detect Product Features
# ===================================================

# TODO: Create three new columns:
#   - is_wireless: TRUE if product name contains "wireless" (case-insensitive)
#   - is_premium: TRUE if product name contains "pro", "premium", or "deluxe"
#   - is_gaming: TRUE if product name contains "gaming" or "gamer"
# Hint: Use str_detect() with str_to_lower() for case-insensitive matching
# Hint: Use | (pipe) in regex for OR conditions

products_clean <- products_clean %>%
  mutate(
    is_wireless = str_detect(str_to_lower(product_name_clean), "wireless"),
    is_premium  = str_detect(str_to_lower(product_name_clean), "pro|premium|deluxe"),
    is_gaming   = str_detect(str_to_lower(product_name_clean), "gaming|gamer")
  )

# ===================================================
# Display Results
# ===================================================
cat("Product Feature Detection:\n")
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# ===================================================
# Summary Statistics
# ===================================================
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless, na.rm = TRUE), "\n")
cat("Premium products:",  sum(products_clean$is_premium,  na.rm = TRUE), "\n")
cat("Gaming products:",   sum(products_clean$is_gaming,   na.rm = TRUE), "\n")


Product Feature Detection:


# A tibble: 10 × 4
   product_name_clean                          is_wireless is_premium is_gaming
   <chr>                                       <lgl>       <lgl>      <lgl>    
 1 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 2 "Samsung Galaxy S23 Ultra 256gb"            FALSE       FALSE      FALSE    
 3 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 4 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 5 "Samsung Galaxy S23 Ultra 256gb"            FALSE       FALSE      FALSE    
 6 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 7 "Dell Xps 13 Laptop - Intel I7 - 16gb Ram"  FALSE       FALSE      FALSE    
 8 "Hp Envy Printer - Wireless - Color"        TRUE        FALSE      FALSE    
 9 "Nike Air Max 270 - Size 10 - Black/White"  FALSE       FALSE      FALSE    
10 "Lg 55\" 4k Smart Tv - Oled Display"        FALSE       FALSE      FALSE    

Feature Summary:
Wir

In [28]:
# ===================================================
# Task 3.2: Extract Product Specifications
# ===================================================
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
    size_number = str_extract(product_name_clean, "\\d+")
  )

# ===================================================
# Display products with extracted sizes
# ===================================================
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()


Extracted Product Specifications:
# A tibble: 10 × 2
   product_name_clean                          size_number
   <chr>                                       <chr>      
 1 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 2 "Samsung Galaxy S23 Ultra 256gb"            23         
 3 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 4 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 5 "Samsung Galaxy S23 Ultra 256gb"            23         
 6 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 7 "Dell Xps 13 Laptop - Intel I7 - 16gb Ram"  13         
 8 "Nike Air Max 270 - Size 10 - Black/White"  270        
 9 "Lg 55\" 4k Smart Tv - Oled Display"        55         
10 "Sony Wh-1000xm4 Wireless Headphones"       1000       


In [30]:
# ===================================================
# Task 3.3: Simple Sentiment Analysis
# ===================================================
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
    positive_words = str_count(feedback_clean, "great|excellent|love|amazing"),
    negative_words = str_count(feedback_clean, "bad|terrible|hate|awful"),
    sentiment_score = positive_words - negative_words
  )

# ===================================================
# Display sentiment analysis results
# ===================================================
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# ===================================================
# Summary
# ===================================================
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score, na.rm = TRUE), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0, na.rm = TRUE), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0, na.rm = TRUE), "\n")


Sentiment Analysis Results:


# A tibble: 10 × 4
   feedback_clean                  positive_words negative_words sentiment_score
   <chr>                                    <int>          <int>           <int>
 1 highly recommend this item                   0              0               0
 2 excellent service                            1              0               1
 3 poor quality control                         0              0               0
 4 average product, nothing speci…              0              0               0
 5 amazing customer support!!!                  1              0               1
 6 amazing customer support!!!                  1              0               1
 7 average product, nothing speci…              0              0               0
 8 good value for money                         0              0               0
 9 highly recommend this item                   0              0               0
10 highly recommend this item                   0              0               0

Overall 

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [32]:
# ================================
# Task 4.1: Parse Transaction Dates
# ================================
library(tidyverse)
library(lubridate)

# We’ll robustly handle your actual column name: "Transaction_DateTime"
# and parse either date-only or datetime strings.

# 1) Detect the date/datetime column
date_col_candidates <- c(
  "transaction_date", "Transaction_Date",
  "transaction_datetime", "Transaction_DateTime",
  "date", "Date", "datetime", "DateTime"
)
tx_col <- intersect(date_col_candidates, names(transactions))[1]
if (is.na(tx_col)) {
  stop("No transaction date/datetime column found. Available columns: ",
       paste(names(transactions), collapse = ", "))
}
cat("Using transactions column for date parsing:", tx_col, "\n")

# 2) Parse to Date (drop time) but accept both date and datetime inputs
#    Try ymd/mdy/dmy with or without times. We pick the first non-NA per row.
transactions_clean <- transactions %>%
  mutate(
    raw_ts = .data[[tx_col]],
    # Try datetime formats first (ymd_hms/mdy_hms/dmy_hms), then date-only (ymd/mdy/dmy)
    parsed_dt = coalesce(
      suppressWarnings(ymd_hms(raw_ts, quiet = TRUE)),
      suppressWarnings(mdy_hms(raw_ts, quiet = TRUE)),
      suppressWarnings(dmy_hms(raw_ts, quiet = TRUE)),
      suppressWarnings(ymd(raw_ts, quiet = TRUE)),
      suppressWarnings(mdy(raw_ts, quiet = TRUE)),
      suppressWarnings(dmy(raw_ts, quiet = TRUE))
    ),
    date_parsed = as_date(parsed_dt)   # final DATE column (no time)
  ) %>%
  select(-raw_ts)

# 3) Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(!!tx_col, date_parsed) %>%
  head(10) %>%
  print()

# (Optional) If you want to continue with components in later tasks:
# transactions_clean <- transactions_clean %>%
#   mutate(
#     year    = year(date_parsed),
#     month   = month(date_parsed),
#     day     = day(date_parsed),
#     wday    = wday(date_parsed, label = TRUE, abbr = TRUE),
#     weekend = wday %in% c("Sat", "Sun"),
#     quarter = quarter(date_parsed),
#     month_name = month(date_parsed, label = TRUE, abbr = FALSE)
#   )


Using transactions column for date parsing: Transaction_DateTime 


Date Parsing Results:
# A tibble: 10 × 2
   Transaction_DateTime date_parsed
   <chr>                <date>     
 1 4/5/24 14:30         NA         
 2 3/15/24 14:30        NA         
 3 3/15/24 14:30        NA         
 4 3/20/24 9:15         NA         
 5 3/20/24 9:15         NA         
 6 3/20/24 9:15         NA         
 7 3/20/24 9:15         NA         
 8 3/15/24 14:30        NA         
 9 25-03-2024 16:45:30  2024-03-25 
10 4/5/24 14:30         NA         


In [33]:
# ===================================================
# Task 4.2: Extract Date Components
# ===================================================
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
    trans_year       = year(date_parsed),
    trans_month      = month(date_parsed),
    trans_month_name = month(date_parsed, label = TRUE, abbr = FALSE),
    trans_day        = day(date_parsed),
    trans_weekday    = wday(date_parsed, label = TRUE, abbr = FALSE),
    trans_quarter    = quarter(date_parsed)
  )

# ===================================================
# Display results
# ===================================================
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_parsed, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()


Date Component Extraction:
# A tibble: 10 × 4
   date_parsed trans_month_name trans_weekday trans_quarter
   <date>      <ord>            <ord>                 <int>
 1 NA          NA               NA                       NA
 2 NA          NA               NA                       NA
 3 NA          NA               NA                       NA
 4 NA          NA               NA                       NA
 5 NA          NA               NA                       NA
 6 NA          NA               NA                       NA
 7 NA          NA               NA                       NA
 8 NA          NA               NA                       NA
 9 2024-03-25  March            Monday                    1
10 NA          NA               NA                       NA


In [34]:
# ===================================================
# Task 4.3: Identify Weekend Transactions
# ===================================================
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
    is_weekend = wday(date_parsed) %in% c(1, 7)
  )

# ===================================================
# Summary
# ===================================================
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend, na.rm = TRUE) / nrow(transactions_clean) * 100, 1), "%\n")


Weekend vs Weekday Transactions:

FALSE 
  150 

Percentage of weekend transactions: 0 %


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [35]:
# ===================================================
# Task 5.1: Calculate Days Since Transaction
# ===================================================
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
  mutate(
    days_since = as.numeric(lubridate::today() - date_parsed)
  )

# ===================================================
# Display results
# ===================================================
cat("Days Since Transaction:\n")
transactions_clean %>%
  select(CustomerID, date_parsed, days_since) %>%   # fallback if customer_name doesn’t exist
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()


Days Since Transaction:
# A tibble: 10 × 3
   CustomerID date_parsed days_since
        <dbl> <date>           <dbl>
 1         31 2024-03-25         565
 2         25 2024-03-25         565
 3         35 2024-03-25         565
 4         12 2024-03-25         565
 5         39 2024-03-25         565
 6          2 2024-03-25         565
 7         11 2024-03-25         565
 8         48 2024-03-25         565
 9         44 2024-03-25         565
10          6 2024-03-25         565


In [36]:
# ===================================================
# Task 5.2: Categorize by Recency
# ===================================================
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
    recency_category = case_when(
      days_since <= 30 ~ "Recent",
      days_since <= 90 ~ "Moderate",
      days_since > 90  ~ "At Risk",
      TRUE             ~ NA_character_
    )
  )

# ===================================================
# Display distribution
# ===================================================
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# ===================================================
# Show At-Risk Customers
# ===================================================
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(CustomerID, date_parsed, days_since) %>%   # fallback if customer_name not present
  arrange(desc(days_since)) %>%
  print()


Recency Category Distribution:

At Risk 
     61 

At-Risk Customers (>90 days):
# A tibble: 61 × 3
   CustomerID date_parsed days_since
        <dbl> <date>           <dbl>
 1         31 2024-03-25         565
 2         25 2024-03-25         565
 3         35 2024-03-25         565
 4         12 2024-03-25         565
 5         39 2024-03-25         565
 6          2 2024-03-25         565
 7         11 2024-03-25         565
 8         48 2024-03-25         565
 9         44 2024-03-25         565
10          6 2024-03-25         565
# ℹ 51 more rows


## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [37]:
# ===================================================
# Task 6.1: Extract First Names and Create Personalized Messages
# ===================================================

library(tidyverse)
library(stringr)

# --- Ensure a customer_name column exists in transactions_clean ---
# 1) Try to find an existing name column on transactions_clean
tx_name_candidates <- c("customer_name", "Customer_Name", "Name")
tx_name_col <- intersect(tx_name_candidates, names(transactions_clean))[1]

if (is.na(tx_name_col)) {
  # 2) Bring a name column from customers via CustomerID
  cust_name_candidates <- c("customer_name", "Customer_Name", "Name")
  cust_name_col <- intersect(cust_name_candidates, names(customers))[1]
  if (is.na(cust_name_col)) {
    stop("Could not find a customer name column in transactions_clean or customers.")
  }
  # Join and standardize to `customer_name`
  transactions_named <- transactions_clean %>%
    left_join(
      customers %>% select(CustomerID, !!sym(cust_name_col)) %>%
        rename(customer_name = !!sym(cust_name_col)),
      by = "CustomerID"
    )
} else {
  # Standardize to `customer_name` if needed
  transactions_named <- transactions_clean %>%
    mutate(customer_name = .data[[tx_name_col]])
}

# --- Build outreach table with first name + personalized message ---
customer_outreach <- transactions_named %>%
  mutate(
    # Extract first token before first space; fall back to whole string if single token
    first_name = str_extract(customer_name %||% "", "^\\w+"),
    first_name = if_else(is.na(first_name) | first_name == "", "Friend", first_name),

    personalized_message = case_when(
      recency_category == "Recent"   ~ paste0("Hi ", first_name, "! Thanks for your recent purchase!"),
      recency_category == "Moderate" ~ paste0("Hi ", first_name, ", we miss you! Check out our new products."),
      recency_category == "At Risk"  ~ paste0("Hi ", first_name, ", it's been a while! Here's a special offer for you."),
      TRUE                           ~ paste0("Hi ", first_name, "!")
    )
  )

# --- Display personalized messages ---
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(customer_name, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()


Personalized Customer Messages:
# A tibble: 10 × 4
   customer_name first_name days_since personalized_message                     
   <chr>         <chr>           <dbl> <chr>                                    
 1 Customer 26   Customer           NA Hi Customer!                             
 2 Customer 21   Customer           NA Hi Customer!                             
 3 Customer 12   Customer           NA Hi Customer!                             
 4 Customer 6    Customer           NA Hi Customer!                             
 5 Customer 32   Customer           NA Hi Customer!                             
 6 Customer 27   Customer           NA Hi Customer!                             
 7 Customer 31   Customer           NA Hi Customer!                             
 8 Customer 30   Customer           NA Hi Customer!                             
 9 Customer 31   Customer          565 Hi Customer, it's been a while! Here's a…
10 Customer 13   Customer           NA Hi Customer!       

In [38]:
# ===================================================
# Task 6.2: Analyze Transaction Patterns by Weekday
# ===================================================
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending

weekday_patterns <- transactions_clean %>%
  group_by(trans_weekday) %>%
  summarise(
    transaction_count = n(),
    total_amount = sum(Amount, na.rm = TRUE),
    avg_amount = mean(Amount, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  arrange(desc(transaction_count))

# ===================================================
# Display results
# ===================================================
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# ===================================================
# Identify busiest day
# ===================================================
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")


Transaction Patterns by Weekday:
# A tibble: 2 × 4
  trans_weekday transaction_count total_amount avg_amount
  <ord>                     <int>        <dbl>      <dbl>
1 NA                           89       22367.       251.
2 Monday                       61       15367.       252.

🔥 Busiest day: NA 


In [39]:
# ===================================================
# Task 6.3: Monthly Transaction Analysis
# ===================================================
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
  group_by(trans_month, trans_month_name) %>%
  summarise(
    transaction_count = n(),
    unique_customers  = n_distinct(CustomerID),
    .groups = "drop"
  ) %>%
  arrange(trans_month)

# ===================================================
# Display results
# ===================================================
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)


Monthly Transaction Patterns:
# A tibble: 3 × 4
  trans_month trans_month_name transaction_count unique_customers
        <dbl> <ord>                        <int>            <int>
1           3 March                           28               22
2           4 April                           33               23
3          NA NA                              89               44


## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [41]:
# Task 7.1: Create Business Intelligence Dashboard
library(tidyverse)
library(lubridate)

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("         BUSINESS INTELLIGENCE SUMMARY\n")
cat(paste(rep("=", 60), collapse = ""), "\n\n", sep = "")

# --------------------------------------------------
# 📦 PRODUCT ANALYSIS
# --------------------------------------------------
cat("📦 PRODUCT ANALYSIS\n")
cat(paste(rep("─", 30), collapse = ""), "\n", sep = "")

# Totals and feature flags
total_products   <- nrow(products_clean)
wireless_count   <- sum(products_clean$is_wireless %||% FALSE, na.rm = TRUE)
premium_count    <- sum(products_clean$is_premium  %||% FALSE, na.rm = TRUE)

# Most common category (using cleaned category if present)
cat_col <- if ("category_clean" %in% names(products_clean)) "category_clean" else
           if ("category" %in% names(products_clean)) "category" else NA_character_
most_common_category <- if (!is.na(cat_col)) {
  products_clean %>% count(.data[[cat_col]], sort = TRUE) %>% slice(1) %>% pull(1)
} else {"(no category column)"}

cat("Total products:", total_products, "\n")
cat("Wireless products:", wireless_count, "\n")
cat("Premium products:", premium_count, "\n")
cat("Most common category:", most_common_category, "\n")

# --------------------------------------------------
# 💬 CUSTOMER SENTIMENT
# --------------------------------------------------
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(paste(rep("─", 30), collapse = ""), "\n", sep = "")

total_feedback <- nrow(feedback_clean)
avg_sentiment  <- mean(feedback_clean$sentiment_score %||% NA_real_, na.rm = TRUE)
pos_pct <- if (total_feedback > 0) {
  round(100 * sum((feedback_clean$sentiment_score %||% 0) > 0, na.rm = TRUE) / total_feedback, 1)
} else { NA_real_ }
neg_pct <- if (total_feedback > 0) {
  round(100 * sum((feedback_clean$sentiment_score %||% 0) < 0, na.rm = TRUE) / total_feedback, 1)
} else { NA_real_ }

cat("Total feedback entries:", total_feedback, "\n")
cat("Average sentiment score:", round(avg_sentiment, 3), "\n")
cat("Percentage positive reviews:", pos_pct, "%\n")
cat("Percentage negative reviews:", neg_pct, "%\n")

# --------------------------------------------------
# 📊 TRANSACTION PATTERNS
# --------------------------------------------------
cat("\n📊 TRANSACTION PATTERNS\n")
cat(paste(rep("─", 30), collapse = ""), "\n", sep = "")

total_tx   <- nrow(transactions_clean)
date_min   <- suppressWarnings(min(transactions_clean$date_parsed, na.rm = TRUE))
date_max   <- suppressWarnings(max(transactions_clean$date_parsed, na.rm = TRUE))

weekday_patterns <- transactions_clean %>%
  group_by(trans_weekday) %>%
  summarise(transaction_count = n(), .groups = "drop") %>%
  arrange(desc(transaction_count))

busiest_day <- if (nrow(weekday_patterns) > 0) as.character(weekday_patterns$trans_weekday[1]) else "(none)"

wknd_pct <- if (total_tx > 0) {
  round(100 * sum(transactions_clean$is_weekend %||% FALSE, na.rm = TRUE) / total_tx, 1)
} else { NA_real_ }

cat("Total transactions:", total_tx, "\n")
cat("Date range:", as.character(date_min), "to", as.character(date_max), "\n")
cat("Busiest weekday:", busiest_day, "\n")
cat("Weekend transaction percentage:", wknd_pct, "%\n")

# --------------------------------------------------
# 👥 CUSTOMER RECENCY
# --------------------------------------------------
cat("\n👥 CUSTOMER RECENCY\n")
cat(paste(rep("─", 30), collapse = ""), "\n", sep = "")

# Choose an identifier for customers
cust_id_col <- intersect(c("CustomerID","customer_id","customerId"), names(transactions_clean))[1]
cust_name_col <- intersect(c("customer_name","Customer_Name","Name"), names(transactions_clean))[1]

# Build a per-customer recency table from last transaction date
if (!is.na(cust_id_col)) {
  recency_by_customer <- transactions_clean %>%
    group_by(.data[[cust_id_col]]) %>%
    summarise(last_date = max(date_parsed, na.rm = TRUE), .groups = "drop") %>%
    mutate(days_since_last = as.numeric(today() - last_date),
           recency_band = case_when(
             days_since_last <= 30 ~ "Recent",
             days_since_last <= 90 ~ "Moderate",
             days_since_last >  90 ~ "At Risk",
             TRUE ~ NA_character_
           ))
  total_customers   <- nrow(recency_by_customer)
  recent_customers  <- sum(recency_by_customer$recency_band == "Recent",  na.rm = TRUE)
  atrisk_customers  <- sum(recency_by_customer$recency_band == "At Risk", na.rm = TRUE)
  reengage_pct      <- if (total_customers > 0) round(100 * atrisk_customers / total_customers, 1) else NA_real_
} else if (!is.na(cust_name_col)) {
  recency_by_customer <- transactions_clean %>%
    group_by(.data[[cust_name_col]]) %>%
    summarise(last_date = max(date_parsed, na.rm = TRUE), .groups = "drop") %>%
    mutate(days_since_last = as.numeric(today() - last_date),
           recency_band = case_when(
             days_since_last <= 30 ~ "Recent",
             days_since_last <= 90 ~ "Moderate",
             days_since_last >  90 ~ "At Risk",
             TRUE ~ NA_character_
           ))
  total_customers   <- nrow(recency_by_customer)
  recent_customers  <- sum(recency_by_customer$recency_band == "Recent",  na.rm = TRUE)
  atrisk_customers  <- sum(recency_by_customer$recency_band == "At Risk", na.rm = TRUE)
  reengage_pct      <- if (total_customers > 0) round(100 * atrisk_customers / total_customers, 1) else NA_real_
} else {
  total_customers <- recent_customers <- atrisk_customers <- reengage_pct <- NA_real_
}

cat("Number of recent customers (≤ 30 days):", recent_customers, "\n")
cat("Number of at-risk customers (> 90 days):", atrisk_customers, "\n")
cat("Percentage needing re-engagement:", reengage_pct, "%\n")


         BUSINESS INTELLIGENCE SUMMARY

📦 PRODUCT ANALYSIS
──────────────────────────────
Total products: 75 
Wireless products: 17 
Premium products: 13 
Most common category: Electronics 

💬 CUSTOMER SENTIMENT
──────────────────────────────
Total feedback entries: 100 
Average sentiment score: 0.18 
Percentage positive reviews: 30 %
Percentage negative reviews: 20 %

📊 TRANSACTION PATTERNS
──────────────────────────────
Total transactions: 150 
Date range: 2024-03-25 to 2024-04-01 
Busiest weekday: NA 
Weekend transaction percentage: 0 %

👥 CUSTOMER RECENCY
──────────────────────────────


Warning message:
“There were 13 warnings in `summarise()`.
The first warning was:
ℹ In argument: `last_date = max(date_parsed, na.rm = TRUE)`.
ℹ In group 4: `CustomerID = 4`.
Caused by warning in `max.default()`:
! no non-missing arguments to max; returning -Inf
ℹ Run `dplyr::last_dplyr_warnings()` to see the 12 remaining warnings.”


Number of recent customers (≤ 30 days): 0 
Number of at-risk customers (> 90 days): 47 
Percentage needing re-engagement: 100 %


In [42]:
# ===================================================
# Task 7.2: Identify Top Products by Category
# ===================================================
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
  group_by(category_clean) %>%
  summarise(
    product_count = n(),
    .groups = "drop"
  ) %>%
  arrange(desc(product_count)) %>%
  slice_head(n = 5)

# ===================================================
# Display results
# ===================================================
cat("Top Product Categories:\n")
print(top_categories)


Top Product Categories:
# A tibble: 5 × 2
  category_clean product_count
  <chr>                  <int>
1 Electronics               21
2 Computers                 15
3 Audio                     14
4 Tv                        14
5 Shoes                     11


## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

Your answer here: Cleaning the text data significantly improved my ability to analyze the data by ensuring consistency across entries. For example, in the product catalog, there were multiple variations of the same product name due to inconsistent casing and extra spaces (e.g., "  Wireless Mouse", "wireless mouse", "WIRELESS MOUSE"). By using functions like `str_trim()` and `str_to_lower()`, I standardized these entries, which allowed me to accurately count occurrences and analyze product popularity without duplication errors.



### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

Your answer here: Detecting patterns in product names provided insights into customer preferences and market trends. For instance, identifying products labeled as "wireless" or "gaming" helped me understand which categories were more popular among customers. A business could use this information to tailor marketing campaigns, stock inventory based on demand, and develop new products that align with customer interests. For example, if "gaming" products are trending, the business could focus on expanding their gaming accessories line to capture more market share.



### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**

Your answer here:

1. Better Inventory Management: By analyzing transaction dates, businesses can identify peak shopping days and months, allowing them to optimize inventory levels and reduce stockouts or overstock situations.
2. Targeted Marketing Campaigns: Understanding when customers are most active enables businesses to schedule promotions and advertising during high-traffic periods, increasing the effectiveness of their marketing efforts.
3. Staffing Optimization: Analyzing transaction patterns helps businesses allocate staff more efficiently during busy periods, ensuring better customer service and reducing wait times.


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

Your answer here: I would recommend the following actions for each customer category: 
- Recent: For recent customers, I would prioritize engagement through personalized thank-you messages and exclusive offers to encourage repeat purchases. These customers are already active, so maintaining their interest is crucial.
- Moderate: For moderate customers, I would suggest targeted promotions and reminders about products they previously viewed
- or purchased. Offering incentives like discounts or loyalty points could help re-engage them and move them back to the recent category.
- At Risk: For at-risk customers, I would recommend a more aggressive re-engagement strategy, such as personalized outreach via email or phone calls. Offering significant discounts or special deals could entice them to return. Prioritizing these actions based on customer lifetime value would ensure that high-value customers receive more attention and resources.
- Overall, I would prioritize actions based on the potential revenue impact, focusing first on recent and high-value customers while still addressing at-risk customers to prevent churn.



### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Your answer here: The sentiment analysis I performed can be used to identify common themes in customer feedback, allowing businesses to address specific issues and improve their products or services. For example, if many customers express dissatisfaction with a particular feature, the company can prioritize enhancements in that area. Additionally, positive feedback can highlight strengths that the business can leverage in marketing campaigns. However, the limitations of this simple sentiment analysis approach include its reliance on keyword detection, which may not capture the full context of customer feedback. For instance, a review that contains both positive and negative sentiments may be misclassified if it only focuses on specific words. Additionally, sarcasm or nuanced language can lead to inaccurate sentiment classification. More advanced techniques, such as machine learning-based sentiment analysis, would provide a deeper understanding of customer opinions.



### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

Your answer here: A real business scenario where combining string manipulation and date analysis would be essential is in managing a subscription-based e-commerce platform. In this scenario, I would need to analyze customer reviews (string data) alongside their subscription renewal dates (date data) to understand customer satisfaction and retention patterns. By cleaning and standardizing the review text, I could extract sentiment and key themes related to product quality, customer service, and overall experience. Simultaneously, analyzing the renewal dates would help identify trends in subscription renewals, such as peak renewal periods or correlations between negative reviews and subscription cancellations. The insights I would be trying to discover include identifying factors that lead to customer churn, understanding the impact of customer feedback on renewal rates, and developing targeted strategies to improve customer satisfaction and retention. This combined analysis would enable the business to make data-driven decisions to enhance both product offerings and customer service strategies.



## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**